gnsi/acltrie/trie.go

```diff
@@ -0,0 +1,199 @@
+// limitations under the License.
+
+// Package acltrie contains a trie for gNMI path with a corresponding policy.
+package acltrie
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+	pathzpb "github.com/openconfig/gnsi/pathz"
+)
+
+// Trie is the root of the ACL trie.
+type Trie struct {
+	root *trieNode
+}
+
+type trieNode struct {
+	// childrenByRank orders the children of the node by number the non-wildcard keys in the child.
+	// This allows fast lookup of the best match: the one with the largest number of non-wildcard keys.
+	childrenByRank []map[string]*trieNode
+	elem           *gpb.PathElem
+	hasPolicy      bool
+	users          policies
+	groups         policies
+}
+
+// policies are map of mode to user to action.
+type policies map[pathzpb.Mode]map[string]pathzpb.Action
+
+// insert adds a principal for a given mode and action to policies, returning an error on duplicate entry.
+func (p policies) insert(principal string, mode pathzpb.Mode, action pathzpb.Action) error {
+	m, ok := p[mode]
+	if !ok {
+		p[mode] = make(map[string]pathzpb.Action)
+		m = p[mode]
+	}
+	if _, ok := m[principal]; ok {
+		return fmt.Errorf("policy already contains action for principal")
+	}
+	m[principal] = action
+
+	return nil
+}
+
+// Insert inserts a new policy into the trie.
+func (t *Trie) Insert(r *pathzpb.AuthorizationRule) error {
+	if t.root == nil {
+		t.root = &trieNode{}
+	}
+	if r.Action == pathzpb.Action_ACTION_UNSPECIFIED {
+		return fmt.Errorf("action unspecified")
+	}
+	if r.Mode == pathzpb.Mode_MODE_UNSPECIFIED {
+		return fmt.Errorf("mode unspecified")
+	}
+	if r.GetGroup() == "" && r.GetUser() == "" {
+		return fmt.Errorf("principal unset")
+	}
+
+	path := r.GetPath()
+	node := t.root
+
+	for i, elem := range path.Elem {
+		if elem.Name == "*" {
+			return fmt.Errorf("wildcard path names are not permitted")
+		}
+
+		// Normalize path string by pruning wildcard keys.
+		pathStr, err := elemToString(elem.Name, elem.Key)
+		if err != nil {
+			return fmt.Errorf("invalid path element: %v", err)
+		}
+
+		// The rank of the child is the number of non-wildcard keys (if any).
+		rank := len(setKeys(elem.Key))
+		if rank >= len(node.childrenByRank) { // Resize the slice to [0, rank].
+			node.childrenByRank = append(node.childrenByRank, make([]map[string]*trieNode, rank-len(node.childrenByRank)+1)...)
+		}
+		if node.childrenByRank[rank] == nil {
+			node.childrenByRank[rank] = make(map[string]*trieNode)
+		}
+		children := node.childrenByRank[rank]
+
+		// If the node already exists, keep going.
+		if _, ok := children[pathStr]; ok {
+			node = children[pathStr]
+			continue
+		}
+
+		// Before adding a new node, check if the path conflicts with another policy.
+		for _, child := range children {
+			if pathElemsMatch(child.elem, elem) {
+				return fmt.Errorf("policy path conflict with %v/%v", path.Elem[0:i], child.elem)
+			}
+		}
```

Comment:
Ok, so what this is saying is that `a[foo][*][*]` doesn't conflict with `a[*][bar][baz]` because the latter has a lower rank and should take precedence? Is that gNSI's intended behaviour?

Reply:
that's right. gNSI spec says to prefer defined key over wildcard element. this implementation works for any number of defined keys.